File: runtimes/unity/Packages/studio.shortsleeve.gamescript/Runtime/Iso4217.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace GameScript
{
    /// <summary>
    /// ISO 4217 currency utilities for locale-aware currency formatting.
    /// Provides minor unit digit counts and currency symbols.
    /// </summary>
    internal static class Iso4217
    {
        // Minor unit digits (decimal places) per ISO 4217 currency code.
        // Only currencies that differ from the default of 2 are listed.
        static readonly Dictionary<string, int> s_MinorUnitOverrides =
            new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase)
        {
            // 0 decimal places
            { "BIF", 0 }, { "BYR", 0 }, { "CLP", 0 }, { "DJF", 0 },
            { "GNF", 0 }, { "ISK", 0 }, { "JPY", 0 }, { "KMF", 0 },
            { "KRW", 0 }, { "PYG", 0 }, { "RWF", 0 }, { "UGX", 0 },
            { "UYI", 0 }, { "VND", 0 }, { "VUV", 0 }, { "XAF", 0 },
            { "XOF", 0 }, { "XPF", 0 },

            // 3 decimal places
            { "BHD", 3 }, { "IQD", 3 }, { "JOD", 3 }, { "KWD", 3 },
            { "LYD", 3 }, { "OMR", 3 }, { "TND", 3 },

            // 4 decimal places
            { "CLF", 4 }, { "UYW", 4 },
        };

        /// <summary>
        /// Returns the number of minor unit digits (decimal places) for the given ISO 4217
        /// currency code. Defaults to 2 for unknown codes (the most common case).
        /// </summary>
        /// <param name="currencyCode">ISO 4217 code (e.g., "USD", "JPY", "KWD").</param>
        /// <returns>Number of decimal places (0, 2, 3, or 4).</returns>
        internal static int GetMinorUnitDigits(string currencyCode)
        {
            if (string.IsNullOrEmpty(currencyCode))
                return 2;

            return s_MinorUnitOverrides.TryGetValue(currencyCode, out int digits) ? digits : 2;
        }

        /// <summary>
        /// Returns the currency symbol for the given ISO 4217 code and culture.
        /// Falls back to the code itself (e.g., "USD") if no symbol is available.
        /// </summary>
        /// <param name="currencyCode">ISO 4217 code (e.g., "USD", "EUR").</param>
        /// <param name="culture">The culture to use for symbol resolution.</param>
        /// <returns>The currency symbol string.</returns>
        internal static string GetSymbol(string currencyCode, CultureInfo culture)
        {
            if (string.IsNullOrEmpty(currencyCode))
                return "";

            // If the culture's own currency matches, use its native symbol
            try
            {
                RegionInfo region = new RegionInfo(culture.Name);
                if (string.Equals(region.ISOCurrencySymbol, currencyCode, StringComparison.OrdinalIgnoreCase))
                    return culture.NumberFormat.CurrencySymbol;
            }
            catch
            {
                // InvariantCulture or unknown culture — fall through
            }

            // For foreign currencies, try to find a culture that uses this currency
            // to get the canonical symbol. Common cases handled inline for performance.
            switch (currencyCode.ToUpperInvariant())
            {
                case "USD": return "$";
                case "EUR": return "€";
                case "GBP": return "£";
                case "JPY": return "¥";
                case "CNY": return "¥";
                case "KRW": return "₩";
                case "INR": return "₹";
                case "RUB": return "₽";
                case "BRL": return "R$";
                case "TRY": return "₺";
                case "THB": return "฿";
                case "PLN": return "zł";
                case "SEK": return "kr";
                case "NOK": return "kr";
                case "DKK": return "kr";
                case "CHF": return "CHF";
                case "CAD": return "CA$";
                case "AUD": return "A$";
                case "NZD": return "NZ$";
                case "MXN": return "MX$";
                case "SGD": return "S$";
                case "HKD": return "HK$";
                case "TWD": return "NT$";
                default:    return currencyCode;
            }
        }
    }
}

File: runtimes/unreal/Source/GameScript/Private/Iso4217.cpp
#include "Iso4217.h"

const TMap<FString, int32>& FIso4217::GetMinorUnitOverrides()
{
	static TMap<FString, int32> Overrides;
	if (Overrides.Num() == 0)
	{
		// 0 decimal places
		Overrides.Add(TEXT("BIF"), 0);
		Overrides.Add(TEXT("BYR"), 0);
		Overrides.Add(TEXT("CLP"), 0);
		Overrides.Add(TEXT("DJF"), 0);
		Overrides.Add(TEXT("GNF"), 0);
		Overrides.Add(TEXT("ISK"), 0);
		Overrides.Add(TEXT("JPY"), 0);
		Overrides.Add(TEXT("KMF"), 0);
		Overrides.Add(TEXT("KRW"), 0);
		Overrides.Add(TEXT("PYG"), 0);
		Overrides.Add(TEXT("RWF"), 0);
		Overrides.Add(TEXT("UGX"), 0);
		Overrides.Add(TEXT("UYI"), 0);
		Overrides.Add(TEXT("VND"), 0);
		Overrides.Add(TEXT("VUV"), 0);
		Overrides.Add(TEXT("XAF"), 0);
		Overrides.Add(TEXT("XOF"), 0);
		Overrides.Add(TEXT("XPF"), 0);

		// 3 decimal places
		Overrides.Add(TEXT("BHD"), 3);
		Overrides.Add(TEXT("IQD"), 3);
		Overrides.Add(TEXT("JOD"), 3);
		Overrides.Add(TEXT("KWD"), 3);
		Overrides.Add(TEXT("LYD"), 3);
		Overrides.Add(TEXT("OMR"), 3);
		Overrides.Add(TEXT("TND"), 3);

		// 4 decimal places
		Overrides.Add(TEXT("CLF"), 4);
		Overrides.Add(TEXT("UYW"), 4);
	}
	return Overrides;
}

int32 FIso4217::GetMinorUnitDigits(const FString& CurrencyCode)
{
	if (CurrencyCode.IsEmpty())
	{
		return 2;
	}

	const FString Upper = CurrencyCode.ToUpper();
	const int32* Found = GetMinorUnitOverrides().Find(Upper);
	return Found ? *Found : 2;
}

FString FIso4217::GetSymbol(const FString& CurrencyCode, const FString& LocaleCode)
{
	if (CurrencyCode.IsEmpty())
	{
		return FString();
	}

	const FString Upper = CurrencyCode.ToUpper();

	// Common currency symbols (matching Unity's fallback table)
	if (Upper == TEXT("USD")) return TEXT("$");
	if (Upper == TEXT("EUR")) return TEXT("\u20AC");
	if (Upper == TEXT("GBP")) return TEXT("\u00A3");
	if (Upper == TEXT("JPY")) return TEXT("\u00A5");
	if (Upper == TEXT("CNY")) return TEXT("\u00A5");
	if (Upper == TEXT("KRW")) return TEXT("\u20A9");
	if (Upper == TEXT("INR")) return TEXT("\u20B9");
	if (Upper == TEXT("RUB")) return TEXT("\u20BD");
	if (Upper == TEXT("BRL")) return TEXT("R$");
	if (Upper == TEXT("TRY")) return TEXT("\u20BA");
	if (Upper == TEXT("THB")) return TEXT("\u0E3F");
	if (Upper == TEXT("PLN")) return TEXT("z\u0142");
	if (Upper == TEXT("SEK")) return TEXT("kr");
	if (Upper == TEXT("NOK")) return TEXT("kr");
	if (Upper == TEXT("DKK")) return TEXT("kr");
	if (Upper == TEXT("CHF")) return TEXT("CHF");
	if (Upper == TEXT("CAD")) return TEXT("CA$");
	if (Upper == TEXT("AUD")) return TEXT("A$");
	if (Upper == TEXT("NZD")) return TEXT("NZ$");
	if (Upper == TEXT("MXN")) return TEXT("MX$");
	if (Upper == TEXT("SGD")) return TEXT("S$");
	if (Upper == TEXT("HKD")) return TEXT("HK$");
	if (Upper == TEXT("TWD")) return TEXT("NT$");

	// Unknown currency — return the code itself
	return CurrencyCode;
}
